spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/runtime/SpekRuntime.kt

```diff
@@ -9,24 +13,54 @@ import org.spekframework.spek2.runtime.execution.ExecutionRequest
 import org.spekframework.spek2.runtime.lifecycle.LifecycleManager
 import org.spekframework.spek2.runtime.scope.*
 import org.spekframework.spek2.runtime.util.ClassUtil
+import kotlin.time.ExperimentalTime
+import kotlin.time.measureTime
 
 class SpekRuntime {
-    fun discover(discoveryRequest: DiscoveryRequest): DiscoveryResult {
-        val scopes = discoveryRequest.context.getTests()
-            .map { testInfo ->
-                val matchingPath = discoveryRequest.paths.firstOrNull { it.intersects(testInfo.path) }
-                testInfo to matchingPath
-            }
-            .filter { (_, matchingPath) -> matchingPath != null }
-            .map { (testInfo, matchingPath) ->
-                checkNotNull(matchingPath)
-                val spec = resolveSpec(testInfo.createInstance(), testInfo.path)
-                spec.filterBy(matchingPath)
-                spec
+    private suspend fun CoroutineScope.filterScopes(discoveryRequest: DiscoveryRequest): Flow<GroupScopeImpl> = flow {
+        val tasks = mutableListOf<Deferred<GroupScopeImpl?>>()
+        discoveryRequest.context.getTests().forEach { testInfo ->
+            val matchingPath = discoveryRequest.paths.firstOrNull { it.intersects(testInfo.path) }
+            if (matchingPath != null) {
+                val task = async {
```

Comment:
Async uses the `CoroutineDispatcher` of its parent, how many threads are backing the said dispatcher will determine if the filtering is done synchronously or asynchronously.